cryostatio/cryostat-operator
# fix(watch): enqueue CR when target namespace is deleted

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Fixes: #941 

## Description of the change:
* Adds labels to role bindings and secrets created by the operator in target namespaces. These labels identify the name and namespace of the CR responsible for them
* Adds a custom controller watch that filters role bindings and secrets by those labels, and enqueues the identified CR
* Adds a testable wrapper around the controller builder, with a fake implementation for tests.
* Adds tests for each controller watch in the `SetupWithManager` method.

## Motivation for the change:
* Allows the operator to watch and respond to deletion of those target namespace objects, or the namespace itself. In the case where the object was deleted, the controller will recreate the deleted object. In the case that the target namespace was deleted, the controller will requeue with an error until either the namespace is recreated, or it is removed from the CR spec.
* Improves test coverage for controller watches, which has been missing.

## How to manually test:
1. Create a multi-namespace Cryostat CR
   a. Delete a role binding and/or secret within the target namespace.
   b. Observe the operator recreating the deleted object.
2. Delete the entire target namespace
   a. The operator should fail to reconcile the CR
   b. Recreate the namespace
   c. The operator should successfully reconcile, and recreate the missing objects
3. Delete the entire target namespace again
   a. The operator should fail to reconcile the CR
   b. Remove the deleted namespace from the CR's spec
   c. The operator should successfully reconcile
